jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/DataUtils.scala

```diff
@@ -16,47 +16,40 @@
 
 package ml.dmlc.xgboost4j.scala.spark
 
-import scala.collection.JavaConverters._
+import ml.dmlc.xgboost4j.{LabeledPoint => XGBLabeledPoint}
 
-import ml.dmlc.xgboost4j.LabeledPoint
-import org.apache.spark.ml.feature.{LabeledPoint => MLLabeledPoint}
-import org.apache.spark.ml.linalg.{DenseVector, SparseVector, Vector}
+import org.apache.spark.ml.linalg.{DenseVector, SparseVector, Vector, Vectors}
 
 object DataUtils extends Serializable {
-
-  implicit def fromSparkPointsToXGBoostPointsJava(sps: Iterator[MLLabeledPoint])
-    : java.util.Iterator[LabeledPoint] = {
-    fromSparkPointsToXGBoostPoints(sps).asJava
-  }
-
-  implicit def fromSparkPointsToXGBoostPoints(sps: Iterator[MLLabeledPoint]):
-      Iterator[LabeledPoint] = {
-    for (p <- sps) yield {
-      p.features match {
-        case denseFeature: DenseVector =>
-          LabeledPoint.fromDenseVector(p.label.toFloat, denseFeature.values.map(_.toFloat))
-        case sparseFeature: SparseVector =>
-          LabeledPoint.fromSparseVector(p.label.toFloat, sparseFeature.indices,
-            sparseFeature.values.map(_.toFloat))
-      }
+  private[spark] implicit class XGBLabeledPointFeatures(
+      val labeledPoint: XGBLabeledPoint
+  ) extends AnyVal {
+    /**
+     * Returns feature of the point as [Vector].
+     *
+     * If the point is sparse, the dimensionality of the resulting sparse
+     * vector would be [[Int.MaxValue]]. This is the only safe value, since
+     * XGBoost does not store the dimensionality explicitly.
+     */
+    def features: Vector = if (labeledPoint.indices == null) {
+      Vectors.dense(labeledPoint.values.map(_.toDouble))
+    } else {
+      Vectors.sparse(Int.MaxValue, labeledPoint.indices, labeledPoint.values.map(_.toDouble))
     }
   }
 
-  implicit def fromSparkVectorToXGBoostPointsJava(sps: Iterator[Vector])
-    : java.util.Iterator[LabeledPoint] = {
-    fromSparkVectorToXGBoostPoints(sps).asJava
-  }
-
-  implicit def fromSparkVectorToXGBoostPoints(sps: Iterator[Vector])
-    : Iterator[LabeledPoint] = {
-    for (p <- sps) yield {
-      p match {
-        case denseFeature: DenseVector =>
-          LabeledPoint.fromDenseVector(0.0f, denseFeature.values.map(_.toFloat))
-        case sparseFeature: SparseVector =>
-          LabeledPoint.fromSparseVector(0.0f, sparseFeature.indices,
-            sparseFeature.values.map(_.toFloat))
-      }
+  private[spark] implicit class MLVectorToXGBLabeledPoint(val v: Vector) extends AnyVal {
```

Comment:
Both extensions are used in the codebase. Moreover, `asXGB` allows to convert `MLLabeledPoint` to `XGBLabeledPoint` in one method call, which is convenient unless we want to duplicate the API (see above).